File: Greedy Algorithms/Fractional Knapsack/cpp/fractional_knapsack.cpp
#include <bits/stdc++.h>

using namespace std;

double get_optimal_value(int capacity, vector<int> weights, vector<int> values) {
  double value = 0.0;
  int n = weights.size();
  for(int i = 1; i < n; i++) {
      int w = weights[i];
      int v = values[i];
      double key = (double)v/w;
      int j = i - 1;
      while(j >= 0 && (double)values[j]/weights[j] < key) {
          weights[j+1] = weights[j];
          values[j+1] = values[j];
          j--;
      }
      weights[j+1] = w;
      values[j+1] = v;
  }
  for(int i = 0; i < n; i++) {
      if(capacity == 0) {
          return value;
      }
      int a = min(weights[i], capacity);
      value += a*((double)values[i]/weights[i]);
      weights[i] -= a;
      capacity -= a;
  }
  return value;
}

int main() {
  int n; int capacity;
  cin >> n >> capacity;
  vector <int> values(n); vector <int> weights(n);
  for (int i = 0; i < n; i++) {
    cin >> values[i] >> weights[i];
  }
  double optimal_value = get_optimal_value(capacity, weights, values);
  cout << optimal_value << endl;
  return 0;
}

File: Greedy Algorithms/Fractional Knapsack/java/FractionalKnapsack.java
import java.util.Arrays;
import java.util.Comparator;

public class FractionalKnapsack {
  
  double getOptimalValue(int capacity, int[] weights, int[] values) {
    double[][] weightValArr = new double[weights.length][2];
    for (int i = 0; i < weightValArr.length; i++) {
      weightValArr[i][0] = weights[i];
      weightValArr[i][1] = values[i];
    }

    Arrays.sort(weightValArr, new Comparator<double[]>() {
      @Override
      public int compare(double[] a, double[] b) {
        double weigthByValue1 = a[0] / a[1];
        double weightByValue2 = b[0] / b[1];
        if (weigthByValue1 == weightByValue2) {
          return 0;
        }
        return weigthByValue1 > weightByValue2 ? 1 : -1;
      }
    });

    double maxValue = 0;
    double weightRemaining = capacity;
    for (int i = 0; i < weightValArr.length; i++) {
      if (weightRemaining == 0) {
        break;
      }
      double weight = weightValArr[i][0];
      double value = weightValArr[i][1];
      if (weightRemaining - weight >= 0) {
        weightRemaining -= weight;
        maxValue += value;
      } else {
        double qty = weightRemaining / weight;
        maxValue += value * qty;
        weightRemaining -= (weight * qty);
      }
    }
    return maxValue;
  }
}
